src/nvim/menu.c

```diff
@@ -1301,11 +1301,15 @@ void ex_emenu(exarg_T *eap)
 
   /* Found the menu, so execute.
    * Use the Insert mode entry when returning to Insert mode. */
-  if (restart_edit
+  if (
+      ((State & INSERT) || restart_edit)
       && !current_SID
       ) {
     mode = (char_u *)"Insert";
     idx = MENU_INDEX_INSERT;
+  } else if (get_real_state() & VISUAL) {
```

Comment:
Shouldn't this extend the condition online 1313, instead of being its own block? There's more logic in that other block which checks for a selection, restores visual mode, etc.